### PR TITLE
ENH running on gpu

### PR DESCRIPTION
**Pull request by tomMoral**

### CODE REVIEW

File: objective.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
 from benchopt import BaseObjective, safe_import_context
 
 with safe_import_context() as import_ctx:
-    from pytorch_lightning import Trainer
     import tensorflow as tf
+    from pytorch_lightning import Trainer
```
Comment (zaccharieramzi) [inline objective.py:5]:
```suggestion
    from pytorch_lightning import Trainer
    import tensorflow as tf

```

Comment (zaccharieramzi) [reply objective.py:5]: Better in the alphabetical order no?

File: objective.py [hunk 2]
```diff
@@ -111,7 +107,7 @@ def compute(self, model):
                 ["train", "test"], [self.tf_dataset, self.tf_test_dataset]
             ):
                 metrics = model.evaluate(
-                    self.tf_dataset.batch(self.batch_size),
+                    dataset.batch(100),
```
Comment (zaccharieramzi) [inline objective.py:110]: wouldn't we want to have this as a settable attribute in the init?

Comment (zaccharieramzi) [reply objective.py:110]: This also would allow to share between the tf and torch computation

File: utils/tf_solver.py [hunk 3]
```diff
@@ -12,31 +14,42 @@
...
             return True, 'Dataset not fit for TF use'
         return False, None
 
-    def set_objective(self, pl_module, trainer, tf_model, tf_dataset):
+    def set_objective(self, pl_module, torch_dataset, tf_model, tf_dataset):
         self.tf_model = tf_model
         self.tf_model.compile(
             optimizer=self.optimizer,
-            loss='categorical_crossentropy',
+            loss='sparse_categorical_crossentropy',
```
Comment (zaccharieramzi) [inline utils/tf_solver.py:39]: Do we know which one is fastest?
I am guessing sparse, but it might be interesting to have in a comment a link specifying it.

File: utils/torch_solver.py [hunk 4]
```diff
@@ -11,13 +13,20 @@
...
+    }
+
+    def set_objective(self, pl_module, torch_dataset, tf_model, tf_dataset):
         self.pl_module = pl_module
-        self.main_trainer = trainer  # we use this in order
-        # to access some elements from the trainer when
-        # initializing it below
+        self.dataloader = torch.utils.data.DataLoader(
+            torch_dataset, batch_size=self.batch_size,
+            num_workers=6
```
Comment (zaccharieramzi) [inline utils/torch_solver.py:28]: we should add a todo here to replace this hard coded value with something more adaptable.

File: utils/torch_solver.py [hunk 5]
```diff
@@ -29,9 +38,10 @@ def run(self, callback):
 
         # Setup the trainer
         trainer = Trainer(
-            max_epochs=-1, callbacks=[BenchoptCallback(callback)]
+            max_epochs=-1, callbacks=[BenchoptCallback(callback)],
+            accelerator="auto", devices=1
```
Comment (zaccharieramzi) [inline utils/torch_solver.py:42]: Here as well we should have a todo to say that we need a better way to set the number of gpus

File: utils/torch_solver.py (hunk 4 above)
Comment (zaccharieramzi) [inline utils/torch_solver.py:28]:
```suggestion
            num_workers=6,  # TODO: replace with e.g. multiprocessing.cpu_count()
```

File: utils/torch_solver.py (hunk 5 above)
Comment (zaccharieramzi) [inline utils/torch_solver.py:42]:
```suggestion
            accelerator="auto", devices=1,  # TODO: replace with e.g. torch.cuda.device_count()
```

**Review by zaccharieramzi — APPROVED** [review]
LGTM, some minor stuff, feel free to correct or not

File: objective.py (hunk 1 above)
Comment (tomMoral) [reply objective.py:5]: I use the length order... ^^

File: objective.py (hunk 2 above)
Comment (tomMoral) [reply objective.py:110]: This is simply for test loss computation so I would not worry too much.
But you are right that we should share this between tf and torch.

File: utils/torch_solver.py (hunk 4 above)
Comment (tomMoral) [reply utils/torch_solver.py:28]: Yes I agree. The `AUTOTUNE` fo `keras` is very powerfull for this and explains that the iterations are mostly faster there....
